### DISCUSSION TITLE
Fatal error: No ObservableObject of type StoreObservableObject...

### DISCUSSION THREAD

**Discussion by artem-krachulov-paysera** [pointfreeco/swift-composable-architecture#2269] [General]
### Description

After upgrading TCA to 0.55.0 I found a very useful [SwichStore update](https://swiftpackageindex.com/pointfreeco/swift-composable-architecture/0.55.0/documentation/composablearchitecture/switchstore) which helps me to remove useless reducers from the SwitchStore. Here is the current code which works with a warning: 

> 'scope(state:)' is deprecated: 'Store.scope' requires an explicit 'action' transform and is intended to be used to transform a store of a parent domain into a store of a child domain.
> 
> When transforming store state into view state, use the 'observe' parameter when constructing a view store.

```
SwitchStore(self.store.scope(state: \.content)) {

  /// useless
  CaseLet(
    state: /SwitchContentState<LoadedState, String>.loading,
    action: Action.noop,
    then: { store in
      WithViewStore(store) { viewStore in
        LoadingView(state: viewStore.state, orientation: .vertical)
      }
    })
  
  CaseLet(
    state: /SwitchContentState<LoadedState, String>.loaded,
    action: Action.main,
    then: LoadedView.init)
  
  /// useless
  CaseLet(
    state: /SwitchContentState<LoadedState, String>.error,
    action: Action.noop,
    then: { store in
      WithViewStore(store) { viewStore in
        ErrorView(text: viewStore.state)
      }
    })
}
```

And updated code:

```
SwitchStore(self.store.scope(state: \.content, action: Action.noop)) { state in
  
  switch state {
  case let .loading(state):
    LoadingView(state: state, orientation: .vertical)
    
  case .loaded:
    CaseLet(
      state: /SwitchContentState<LoadedState, String>.loaded,
      action: Action.loaded,
      then: LoadedView.init)
    
  case let .error(text):
    ErrorView(text: text)
    
  case .ready:
    EmptyView()
  }
}
```
Only this code triggers an error after run:

> SwiftUI/EnvironmentObject.swift:82: Fatal error: No ObservableObject of type StoreObservableObject<SwitchContentState<LoadedState, String>, Action> found. A View.environmentObject(_:) for StoreObservableObject<SwitchContentState<LoadedState, String>, Action> may be missing as an ancestor of this view.

What can it be? Any way to solve it?

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [ ] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### The Composable Architecture version information

0.55.0

### Destination operating system

iOS 16.5.1

### Xcode version information

Version 14.3.1 (14E300c)

### Swift Compiler version information

```shell
swift-driver version: 1.75.2 Apple Swift version 5.8.1 (swiftlang-5.8.0.124.5 clang-1403.0.22.11.100)
Target: x86_64-apple-macosx13.0
```

**Comment by mbrandonw** [comment]
The `SwitchStore` should be providing the environment object to the body. Are you able to reproduce this in a simple example that we can run to see what is going on?

I would also like to draw your attention to the "Important" note in the docs of `SwitchStore`:

> **Important**
The SwitchStore view builder is only evaluated when the case of state passed to it changes. As such, you should not rely on this value for anything other than checking the current case, e.g. by switching on it and routing to an appropriate CaseLet.

This means for this code:

```swift
case let .loading(state):
  LoadingView(state: state, orientation: .vertical)
…    
case let .error(text):
  ErrorView(text: text)
```

…the views will only be rendered a single type when the case switches to `loading` or `error`. In particular, if the state inside the associated value changes it will not cause the views to re-render. That may be completely fine for your use case, but just want to call it out.

**Comment by artem-krachulov-paysera** [comment]
@mbrandonw Thank you for your response. Here is a demo project with two cases
[SwitchStoreTest.zip](https://github.com/pointfreeco/swift-composable-architecture/files/11977268/SwitchStoreTest.zip)

Also when reproducing, found something. When you remove the action from:

`SwitchStore(self.store.scope(state: \.content, action: Updated.Action.noop)) { state in`

crash solves but a warning appears:

> 'scope(state:)' is deprecated: 'Store.scope' requires an explicit 'action' transform and is intended to be used to transform a store of a parent domain into a store of a child domain.
When transforming store state into view state, use the 'observe' parameter when constructing a view store.

**Comment by sehdich** [comment]
Interesting - it also works without any crash when using an `action` of `{ $0 }`:

`SwitchStore(self.store.scope(state: \.content, action: { $0 })) { ...}`

**Comment by mbrandonw** [comment]
Hi @artemkrachulov, thanks for the repro and now I see what is going on. The problem is that your types do not match up. You are "switching" on a store that has `Never` for its actions, yet you are trying to extract a "case let" store from it that has actions of `Content.Action`. That can't possibly work.

And the reason it works with the "old" version of your demo is because you are not scoping the main store's actions down to `Never`. So, the way to get your demo working is to leave the actions intact with an `action: { $0 }` argument.

It would of course be best if we could catch these things at compile time instead of runtime, but often that is not possible in SwiftUI. Even vanilla SwiftUI has similar problems, such as no guarantees that the value you use in `.tag(…)` matches the kind of binding you provide to `TabView`.

Since this is not an issue in the library I am going to convert it to a discussion.

**Comment by artem-krachulov-paysera** [comment]
@mbrandonw Oh yes, when I started using the switch content action, everything fell into place.

```
public enum SwitchContentAction<T: Equatable>: Equatable {
  case ready
  case loading(Never)
  case loaded(T)
  case error(Never)
}

```
Now It has followed look:

```
SwitchStore(self.store.scope(state: \.content, action: Action.content)) { state in
  switch state {
  case let .loading(text):
    Text(text: text)
    
  case .loaded:
    CaseLet(
      state: /SwitchContentState<LoadedState>.loaded,
      action: SwitchContentAction<LoadedAction>.loaded,
      then: LoadedView.init)
    
  case let .error(text):
    Text(text: text)
    
  case .ready:
    EmptyView()
  }
}
```

Thx
